vmm/src/migration/mod.rs

```diff
@@ -0,0 +1,88 @@
+// Copyright © 2019 Intel Corporation
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+use std::{io, result, thread};
+use std::sync::mpsc::{channel, Sender, SendError, RecvError};
+use crate::api::ApiResponse;
+use crate::migration::state::MigrationState;
+
+pub mod state;
+
+/// Errors associated with VM management
+#[derive(Debug)]
+pub enum Error {
+    /// Cannot spawn a new migration thread.
+    MigrationSpawn(io::Error),
+
+    /// Request send error
+    RequestSend(SendError<MigrationRequest>),
+
+    /// Request receive error
+    RequestRecv(RecvError),
+}
+pub type Result<T> = result::Result<T, Error>;
+
+#[derive(Debug)]
+pub enum MigrationRequest {
+    /// Start to take snapshot, i.e. save Vm states
+    TakeSnapshot,
+
+    /// Start to restore snapshot, i.e. restore Vm states
+    RestoreSnapshot,
+}
+
+#[derive(Clone)]
+pub struct Migration {
+    state: MigrationState,
+    sender: Sender<MigrationRequest>,
+}
+
+impl Migration {
+    pub fn new() -> Result<Self> {
+        let (sender, receiver) = channel();
+        let state = MigrationState::new();
+        let state1 = state.clone();
+
+        thread::Builder::new()
+            .name(format!("migration"))
+            .spawn(move || {
+                let request = receiver.recv().map_err(Error::RequestRecv);
+
+                match request {
+                    Ok(req) => {
+                        match req {
+                            MigrationRequest::TakeSnapshot => {
+                                state1.get_iter().expect("Fail to get migration states");
+                            }
+                            MigrationRequest::RestoreSnapshot => {
+                                /* TODO: call state's restore interface */
+                            }
```

Comment:
I think there is one possibility: we may restore snapshot on a broken (or exceptional) VM, like restore action on Windows. So I left this request here to discuss with you.

Reply:
So that would be kind of a "reset and restore"? Basically having a possibility to reinitialize an already running guest, reset it and restore it from a known good state (a snapshot)?
That's definitely an interesting you case, but could we keep it as a follow up feature? I'm really trying to keep things simple at first :-)

Reply:
Ok, I will add it in todo list for next phase. Will remove it in next formal patch set.